File: widgetssdk/src/main/java/com/glia/widgets/filepreview/domain/usecase/GetImageFileFromNetworkUseCase.java
package com.glia.widgets.filepreview.domain.usecase;

import android.graphics.Bitmap;

import com.glia.androidsdk.chat.AttachmentFile;
import com.glia.widgets.chat.helper.FileHelper;
import com.glia.widgets.filepreview.data.GliaFileRepository;
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException;
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException;

import io.reactivex.Maybe;

public class GetImageFileFromNetworkUseCase {
    private final GliaFileRepository gliaFileRepository;

    public GetImageFileFromNetworkUseCase(GliaFileRepository gliaFileRepository) {
        this.gliaFileRepository = gliaFileRepository;
    }

    public Maybe<Bitmap> execute(AttachmentFile file) {
        if (file == null || file.getName().isEmpty()) {
            return Maybe.error(new FileNameMissingException());
        }
        if (file.isDeleted()) {
            return Maybe.error(new RemoteFileIsDeletedException());
        }

        return gliaFileRepository
                .loadImageFileFromNetwork(file)
                .flatMap(bitmap ->
                        gliaFileRepository
                                .putImageToCache(FileHelper.getFileName(file), bitmap)
                                .andThen(Maybe.just(bitmap))
                );
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/filepreview/domain/usecase/GetImageFileFromNetworkUseCase.kt
package com.glia.widgets.filepreview.domain.usecase

import android.graphics.Bitmap
import com.glia.androidsdk.chat.AttachmentFile
import com.glia.widgets.chat.domain.DecodeSampledBitmapFromInputStreamUseCase
import com.glia.widgets.filepreview.data.GliaFileRepository
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException
import com.glia.widgets.helper.fileName
import io.reactivex.rxjava3.core.Maybe

internal class GetImageFileFromNetworkUseCase(
    private val gliaFileRepository: GliaFileRepository,
    private val decodeSampledBitmapFromInputStreamUseCase: DecodeSampledBitmapFromInputStreamUseCase,
) {
    operator fun invoke(file: AttachmentFile?): Maybe<Bitmap> = when {
        file?.name.isNullOrBlank() -> Maybe.error(FileNameMissingException())
        file!!.isDeleted -> Maybe.error(RemoteFileIsDeletedException())
        else -> gliaFileRepository.loadImageFileFromNetwork(file)
            .flatMap { decodeSampledBitmapFromInputStreamUseCase(it) }
            .doOnSuccess { gliaFileRepository.putImageToCache(file.fileName, it) }
    }
}
